google/resource_cloudfunctions_function.go

```diff
@@ -0,0 +1,462 @@
+
+	if v, ok := d.GetOk("memory"); ok {
+		memory := v.(int)
+		if FUNCTION_ALLOWED_MEMORY[memory] != true {
+			return fmt.Errorf("Allowed values for memory are: 128MB, 256MB, 512MB, 1024MB, and 2048MB. "+
+				"Got %d", memory)
+		}
+		function.AvailableMemoryMb = int64(memory)
+	}
+
+	if v, ok := d.GetOk("description"); ok {
+		function.Description = v.(string)
+	}
+
+	if v, ok := d.GetOk("entry_point"); ok {
+		function.EntryPoint = v.(string)
+	}
+
+	if v, ok := d.GetOk("source"); ok {
+		function.SourceArchiveUrl = v.(string)
+	}
+
+	if v, ok := d.GetOk("timeout"); ok {
+		function.Timeout = fmt.Sprintf("%vs", v.(int))
+	}
+
+	v, triggHttpOk := d.GetOk("trigger_http")
+	if triggHttpOk {
+		if v.(bool) == true {
+			function.HttpsTrigger = &cloudfunctions.HttpsTrigger{
+				Url: "",
+			}
+		}
+	}
+
+	v, triggTopicOk := d.GetOk("trigger_topic")
+	if triggTopicOk {
+		//Make PubSub event publish as in https://cloud.google.com/functions/docs/calling/pubsub
+		function.EventTrigger = &cloudfunctions.EventTrigger{
+			//Other events are not supported
+			EventType: "providers/cloud.pubsub/eventTypes/topic.publish",
+			//Must be like projects/PROJECT_ID/topics/NAME
+			//Topic must be in same project as function
+			Resource: fmt.Sprintf("projects/%s/topics/%s", project, v.(string)),
+		}
+	}
+
+	v, triggBucketOk := d.GetOk("trigger_bucket")
+	if triggBucketOk {
+		//Make Storage event as in https://cloud.google.com/functions/docs/calling/storage
+		function.EventTrigger = &cloudfunctions.EventTrigger{
+			EventType: "providers/cloud.storage/eventTypes/object.change",
+			//Must be like projects/PROJECT_ID/buckets/NAME
+			//Bucket must be in same project as function
+			Resource: fmt.Sprintf("projects/%s/buckets/%s", project, v.(string)),
+		}
+	}
+
+	if !triggHttpOk && !triggTopicOk && !triggBucketOk {
+		return fmt.Errorf("One of aguments [trigger_topic, trigger_bucket, trigger_http] is required: " +
+			"You must specify a trigger when deploying a new function.")
+	}
+
+	if _, ok := d.GetOk("labels"); ok {
+		function.Labels = expandLabels(d)
+	}
+
+	log.Printf("[DEBUG] Creating cloud function: %s", function.Name)
+	op, err := service.Projects.Locations.Functions.Create(
+		createCloudFunctionsPathString(CLOUDFUNCTIONS_REGION_ONLY, project, region, ""), function).Do()
+	if err != nil {
+		return err
+	}
+	err = cloudFunctionsOperationWait(config.clientCloudFunctions, op, "Creating CloudFunctions Function")
+	if err != nil {
+		return err
+	}
+
+	//Name of function should be unique
+	d.SetId(d.Get("name").(string))
+
+	return resourceCloudFunctionsRead(d, meta)
+}
+
+func resourceCloudFunctionsRead(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+
+	service := config.clientCloudFunctions
+
+	project, err := getProject(d, config)
+	if err != nil {
+		return err
+	}
+
+	region, err := getRegion(d, config)
+	if err != nil {
+		return err
+	}
+
+	name := d.Get("name").(string)
```

Comment:
If you use d.Id() here instead of d.Get("name").(string), then you can use an `ImportStatePassthrough` importer instead of a custom one

Reply:
Done :)